presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java

```diff
@@ -1165,6 +1166,11 @@ public SystemSessionProperties(
                         "Enable streaming for partial aggregation",
                         featuresConfig.isStreamingForPartialAggregationEnabled(),
                         false),
+                booleanProperty(
+                        PREFER_MERGE_JOIN,
+                        "Prefer merge join for sorted join inputs, e.g., tables pre-sorted, pre-partitioned by join columns",
```

Comment:
order_based_execution_enabled must be enabled to make merge join planning work.

Reply:
I'm not sure if we should mention that given `order_based_execution_enabled ` would be a connector session property. And from engine level, it shouldn't indicate details on how connector implements its feature.

Thus I put the abstract and common requirement here: sorted join inputs

Reply:
I think we need to document somewhere about what are all the session properties required to enable this feature. Otherwise, other developers/users will have a hard time to figure it out. This seems to be a good place to add this documentation as this session property name sounds like it will enable merge join. If you have a better place to document this, that is also fine.

> And from engine level, it shouldn't indicate details on how connector implements its feature.

For code/design perspective, I definitely agree to have a good abstraction. But for comments, I think it is better to document all the assumptions. E.g., we could add note here to say that currently this is only supported for pre-sorted hive tables and requires enabling order_based_execution_enabled to make it work.

Reply:
Why do we need an extra property order_based_execution_enabled?  Can we always use order based execution if it's supported by the table and prefer_merge_join is true.

Also, if there's no execution support yet for this property we should set it to hidden.

Reply:
I don't think setting it to "hidden" will prevent you from setting it in a cpp cluster.  It's more about revealing it to users.  It means it doesn't show up if you say "show session".  

I think it's totally fine to have one property for streaming aggregation and one for streaming join.  I'm just thinking in general that features are easier to use if you only have to set one property to turn them on.  Otherwise, it's easy to have mistakes where someone sets one and thinks they've enabled the feature because they don't realize you need to set both. Having two properties to enable two different features is easier to reason about as long as the names make it clear what they do.(e.g. `prefer_streaming_aggregation` and `prefer_streaming_join`). 

That being said if you would expect users to want to turn on/off streaming stuff all in one go, then that could be a reason to introduce the master property.  e.g. for spill we have the `spill_enabled` property, since in general you either want spill on or off, but we also have properties for disabling spill for individual operators in case there are bugs or issues with specific operator implementations. So for this case you could have one property `prefer_streaming_operators` and then other properties, `streaming_aggregation_enabled` and `streaming_join_enabled`.  I think a key there is having the operator specific knobs default to true once there are tested java implementations so that users mostly still just have to think about one property.  I still think just having separate properties is simpler though

What do you think?

Reply:
Setting the property to be hidden
Adding the description on order_based_execution_enabled as well

Reply:
@rschlussel 
The reason why we can't always enable order_based_execution_enabled is that it would cause performance regression without working together with merge join or streaming aggregation

Reply:
I apologize, I think I misunderstood the situation before.  Am I understanding correctly as follows: `order_based_execution_enabled` is a property of the hive connector that determines whether to reveal the underlying sortedness of the table to the optimizer via the TableLayout. `prefer_merge_join` is a flag to enable/disable the merge join optimizer.  That seems totally fine to me.